File: app/macos/NativeStream/NativeStream/Views/Player/PlayerSidebar.swift
// PlayerSidebar.swift

import SwiftUI


// MARK: - Player Sidebar

struct PlayerSidebar: View {

    @Environment(EPGViewModel.self)      private var epgVM
    @Environment(PlaylistViewModel.self) private var playlistVM
    @Environment(PlayerViewModel.self)   private var playerVM

    let currentChannel: Channel?

    enum SidebarTab { case onNow, schedule }
    @State private var tab: SidebarTab = .onNow

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 0) {
                sidebarTab("On now",   tab: .onNow)
                sidebarTab("Schedule", tab: .schedule)
            }
            .background(Color.black.opacity(0.6))
            .overlay(alignment: .bottom) {
                Rectangle().fill(Color.white.opacity(0.07)).frame(height: 0.5)
            }

            switch tab {
            case .onNow:    PlayerOnNowTab(currentChannel: currentChannel)
            case .schedule: PlayerScheduleTab(channel: playerVM.currentChannel)
            }
        }
        .frame(width: NS.Player.sidebarWidth)
        .background(Color(hex: "0e0e0e"))
        .overlay(alignment: .leading) {
            Rectangle().fill(Color.white.opacity(0.07)).frame(width: 0.5)
        }
    }

    private func sidebarTab(_ label: String, tab: SidebarTab) -> some View {
        Button { self.tab = tab } label: {
            Text(label)
                .font(NS.Font.captionMed)
                .foregroundStyle(self.tab == tab ? Color.white.opacity(0.85) : Color.white.opacity(0.35))
                .frame(maxWidth: .infinity)
                .padding(.vertical, NS.Spacing.sm)
                .overlay(alignment: .bottom) {
                    if self.tab == tab {
                        Rectangle().fill(NS.accent).frame(height: 1.5)
                    }
                }
        }
        .buttonStyle(.plain)
    }
}


// MARK: - On Now Tab

struct PlayerOnNowTab: View {

    @Environment(EPGViewModel.self)      private var epgVM
    @Environment(PlaylistViewModel.self) private var playlistVM
    @Environment(PlayerViewModel.self)   private var playerVM

    let currentChannel: Channel?

    // Only channels with live or upcoming EPG content.
    // Falls back to all channels if EPG has no data at all.

    private var filteredChannels: [Channel] {
        let all = playlistVM.channels
        let withEPG = all.filter {
            epgVM.currentProgramme(for: $0) != nil || epgVM.nextProgramme(for: $0) != nil
        }
        return withEPG.isEmpty ? all : withEPG
    }

    private var sortedChannels: [Channel] {
        filteredChannels.sorted { a, b in
            let aPlaying = playerVM.currentChannel?.id == a.id
            let bPlaying = playerVM.currentChannel?.id == b.id
            if aPlaying != bPlaying { return aPlaying }
            let aLive = epgVM.currentProgramme(for: a) != nil
            let bLive = epgVM.currentProgramme(for: b) != nil
            if aLive != bLive { return aLive }
            return epgVM.nextProgramme(for: a) != nil && epgVM.nextProgramme(for: b) == nil
        }
    }

    var body: some View {
        ScrollView {
            LazyVStack(spacing: 2) {
                ForEach(sortedChannels) { channel in
                    PlayerSidebarRow(channel: channel)
                }
            }
            .padding(NS.Spacing.sm)
        }
        .onAppear {
            if playerVM.channelList.isEmpty {
                playerVM.channelList = sortedChannels
            }
        }
    }
}

// MARK: - Schedule Tab

struct PlayerScheduleTab: View {

    @Environment(EPGViewModel.self)    private var epgVM
    @Environment(PlayerViewModel.self) private var playerVM

    let channel: Channel?

    private var programmes: [Programme] {
        guard let ch = channel else { return [] }
        return epgVM.schedule(for: ch, hours: 12).sorted { $0.start < $1.start }
    }

    var body: some View {
        ScrollView {
            LazyVStack(spacing: 2) {
                ForEach(programmes, id: \.id) { prog in
                    
                    scheduleRow(prog)
                }
            }
            .padding(NS.Spacing.sm)
        }
    }

    private func scheduleRow(_ prog: Programme) -> some View {
        VStack(alignment: .leading, spacing: NS.Spacing.xs) {
            Text(prog.startTimeString)
                .font(NS.Font.monoSm)
                .foregroundStyle(prog.isNow ? NS.accent : Color.white.opacity(0.3))
            if prog.isNow {
                Text("Now playing")
                    .font(.system(size: NS.Schedule.microLabelSize, weight: .bold))
                    .foregroundStyle(NS.accent)
                    .kerning(0.5)
            }
            Text(prog.title)
                .font(NS.Font.captionMed)
                .foregroundStyle(prog.isNow ? Color.white.opacity(0.85) : Color.white.opacity(0.4))
                .lineLimit(2)
            if prog.isNow {
                NSProgressBar(value: prog.progress, height: 2, glow: false).opacity(0.6)
            }
        }
        .padding(NS.Spacing.sm)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(prog.isNow ? NS.accentGlow : Color.clear)
        .clipShape(RoundedRectangle(cornerRadius: NS.Radius.md))
        .overlay(
            RoundedRectangle(cornerRadius: NS.Radius.md)
                .stroke(prog.isNow ? NS.accentBorder : Color.clear, lineWidth: 0.5)
        )
        .opacity(prog.stop < Date() ? 0.4 : 1.0)
    }
}

File: app/android/app/src/main/java/com/nativestream/android/ui/screens/player/PlayerSidebar.kt
// app/src/main/java/com/nativestream/android/ui/screens/player/PlayerSidebar.kt
//
// Player Sidebar
// 240dp collapsible sidebar with two tabs: On Now and Schedule.
// Channel switch via ExoPlayer.setMediaItem() through PlayerViewModel.play().

package com.nativestream.android.ui.screens.player

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.slideInHorizontally
import androidx.compose.animation.slideOutHorizontally
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.nativestream.android.domain.model.Channel
import com.nativestream.android.domain.model.Programme
import com.nativestream.android.ui.components.NSProgressBar
import com.nativestream.android.ui.theme.NSColors
import com.nativestream.android.ui.theme.NSDimens
import com.nativestream.android.ui.theme.NSType
import com.nativestream.android.ui.viewmodel.EpgViewModel
import com.nativestream.android.ui.viewmodel.PlayerViewModel

private val TAB_INDICATOR_H    = 1.5.dp
private val SCHEDULE_HOURS     = 12

enum class SidebarTab { ON_NOW, SCHEDULE }

@Composable
fun PlayerSidebarContent(
    playerViewModel: PlayerViewModel,
    epgViewModel: EpgViewModel,
    modifier: Modifier = Modifier,
) {
    val activeChannel by playerViewModel.activeChannel.collectAsState()
    var selectedTab   by remember { mutableStateOf(SidebarTab.ON_NOW) }

    Box(
        modifier = modifier
            .width(NSDimens.current.player.sidebarWidth)  // 230dp token
            .fillMaxHeight()
            .background(Color(0xFF0E0E0E))
            .border(
                width = 0.5.dp,
                color = Color.White.copy(alpha = 0.07f),
                shape = RoundedCornerShape(0.dp),
            ),
    ) {
        Column {
            SidebarTabBar(selectedTab = selectedTab, onSelectTab = { selectedTab = it })
            when (selectedTab) {
                SidebarTab.ON_NOW   -> OnNowTab(
                    currentChannel  = activeChannel,
                    playerViewModel = playerViewModel,
                    epgViewModel    = epgViewModel,
                )
                SidebarTab.SCHEDULE -> ScheduleTab(
                    channel      = activeChannel,
                    epgViewModel = epgViewModel,
                )
            }
        }
    }
}

@Composable
fun PlayerSidebar(
    isVisible: Boolean,
    playerViewModel: PlayerViewModel,
    epgViewModel: EpgViewModel,
    modifier: Modifier = Modifier,
) {
    AnimatedVisibility(
        visible  = isVisible,
        enter    = slideInHorizontally { it },
        exit     = slideOutHorizontally { it },
        modifier = modifier,
    ) {
        PlayerSidebarContent(
            playerViewModel = playerViewModel,
            epgViewModel    = epgViewModel,
        )
    }
}

// ── Tab bar ───────────────────────────────────────────────────────────────────

@Composable
private fun SidebarTabBar(selectedTab: SidebarTab, onSelectTab: (SidebarTab) -> Unit) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .background(Color.Black.copy(alpha = 0.6f)),
    ) {
        Row {
            SidebarTab.entries.forEach { tab ->
                val isActive = selectedTab == tab
                val label    = if (tab == SidebarTab.ON_NOW) "On now" else "Schedule"
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    modifier = Modifier
                        .weight(1f)
                        .clickable { onSelectTab(tab) }
                        .padding(vertical = NSDimens.current.spacing.sm),
                ) {
                    Text(
                        text  = label,
                        style = NSType.captionMedium(),
                        color = Color.White.copy(alpha = if (isActive) 0.85f else 0.35f),
                    )
                    if (isActive) {
                        Spacer(modifier = Modifier.height(NSDimens.current.spacing.xs))
                        Box(
                            modifier = Modifier
                                .fillMaxWidth()
                                .height(TAB_INDICATOR_H)
                                .background(NSColors.accent),
                        )
                    }
                }
            }
        }
        // Bottom border
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(0.5.dp)
                .background(Color.White.copy(alpha = 0.07f))
                .align(Alignment.BottomCenter),
        )
    }
}

// ── On Now tab ────────────────────────────────────────────────────────────────

@Composable
private fun OnNowTab(
    currentChannel: Channel?,
    playerViewModel: PlayerViewModel,
    epgViewModel: EpgViewModel,
) {
    val allChannels by playerViewModel.channelList.collectAsState()

    val filteredChannels = remember(allChannels) {
        val withEpg = allChannels.filter {
            epgViewModel.currentProgramme(it) != null || epgViewModel.nextProgramme(it) != null
        }
        withEpg.ifEmpty { allChannels }
    }

    val sortedChannels = remember(filteredChannels) {
        filteredChannels.sortedWith(compareBy(
            { epgViewModel.currentProgramme(it) == null },
            { epgViewModel.nextProgramme(it) == null },
        ))
    }
    val listState = rememberLazyListState()
    val activeIndex = remember(sortedChannels, currentChannel) {
        sortedChannels.indexOfFirst { it.id == currentChannel?.id }
    }

    LaunchedEffect(activeIndex) {
        if (activeIndex >= 0) {
            val visibleItems = listState.layoutInfo.visibleItemsInfo
            val isVisible = visibleItems.any { it.index == activeIndex }
            if (!isVisible) listState.scrollToItem(activeIndex)
        }
    }

    LazyColumn(
        state = listState,
        modifier = Modifier
            .fillMaxHeight()
            .padding(NSDimens.current.spacing.sm)
    ) {
        items(sortedChannels, key = { it.id }) { channel ->
            PlayerSidebarRow(
                channel         = channel,
                playerViewModel = playerViewModel,
                epgViewModel    = epgViewModel,
            )
        }
    }
}

// ── Schedule tab ──────────────────────────────────────────────────────────────

@Composable
private fun ScheduleTab(channel: Channel?, epgViewModel: EpgViewModel) {
    val dimens = NSDimens.current

    val programmes: List<Programme> = remember(channel) {
        channel?.let {
            epgViewModel.schedule(it, hours = SCHEDULE_HOURS)
                .sortedBy { prog -> prog.startEpochMs }
        } ?: emptyList()
    }

    LazyColumn(modifier = Modifier.fillMaxHeight().padding(dimens.spacing.sm)) {
        items(programmes, key = { it.id }) { prog ->
            ScheduleRow(programme = prog)
        }
    }
}

@Composable
private fun ScheduleRow(programme: Programme) {
    val dimens  = NSDimens.current
    val isPast  = programme.stopEpochMs < System.currentTimeMillis()

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(dimens.radius.md))
            .background(if (programme.isNow) NSColors.accentGlow else Color.Transparent)
            .border(
                0.5.dp,
                if (programme.isNow) NSColors.accentBorder else Color.Transparent,
                RoundedCornerShape(dimens.radius.md),
            )
            .padding(dimens.spacing.sm)
            .then(if (isPast) Modifier.then(Modifier.padding(0.dp)) else Modifier),
        ) {
        Text(
            text  = programme.startTimeString,
            style = NSType.monoSmall(),
            color = if (programme.isNow) NSColors.accent else Color.White.copy(alpha = 0.3f),
        )
        if (programme.isNow) {
            Text(
                text  = "Now playing",
                style = NSType.monoSmall(),
                color = NSColors.accent,
            )
        }
        Text(
            text     = programme.title,
            style    = NSType.captionMedium(),
            color    = Color.White.copy(alpha = if (programme.isNow) 0.85f else 0.4f),
            maxLines = 2,
        )
        if (programme.isNow) {
            NSProgressBar(
                value    = programme.progress.toFloat(),
                modifier = Modifier.padding(top = dimens.spacing.xs),
            )
        }
    }
}